src/StatementDumper.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Helmich\TypoScriptParser;
+
+use Helmich\TypoScriptParser\Parser\AST\ConditionalStatementTerminator;
+use Helmich\TypoScriptParser\Parser\AST\NopStatement;
+use Helmich\TypoScriptParser\Parser\AST\Statement;
+
+class StatementDumper
+{
+    private bool $dumpLineNumbers;
+
+    /**
+     * @param array{dumpLineNumbers?: bool} $options
+     */
+    public function __construct(array $options = [])
+    {
+        $this->dumpLineNumbers = $options['dumpLineNumbers'] ?? false;
+    }
+
+    /**
+     * @param array<Statement> $statements
+     */
+    public function dump(array $statements): string
+    {
+        return $this->dumpRecursive($statements);
+    }
+
+    private function dumpRecursive(mixed $statement): string
+    {
+        if ($statement instanceof NopStatement) {
+            return get_class($statement) . '()';
+        }
+
+        if ($statement instanceof Node) {
+            $r = get_class($statement);
+            if ($this->dumpLineNumbers && $lineNumber = $this->dumpLineNumber($statement)) {
+                $r .= $lineNumber;
+            }
+            $r .= '(';
+
+            $properties = $statement->getSubNodeNames();
+            foreach ($properties as $key) {
+                $r .= "\n    " . $key . ': ';
+
+                $value = $statement->$key;
+                if ($value === null) {
+                    $r .= 'null';
+                } elseif ($value === true) {
+                    $r .= 'true';
+                } elseif ($value === false) {
+                    $r .= 'false';
+                } elseif (is_scalar($value)) {
+                    $r = rtrim($r . $value);
+                } else {
+                    $r .= str_replace("\n", "\n    ", $this->dumpRecursive($value));
+                }
```

Comment:
This part and lines 64-76 seem duplicated. Can we extract them into a function?

Reply:
This is a bit complicated because in one case I'm overwriting $r while the others are appended.

Reply:
This is a bit complicated because in one case I'm overwriting $r while the others are appended.